use-cases/core/median-finding/median_finding.cpp

```diff
@@ -0,0 +1,186 @@
+    // evenly spaced, then squared, so half end up being below 0.25
+    double cut_line = 0.5;
+
+    // just for logging, get the time on the main loop
+    // currently I'm interested to see how n_teams impacts the total time
+    typedef std::chrono::high_resolution_clock Clock;
+    auto clock_start = Clock::now();
+
+    // main loop will reduce over threads (inner) and then teams (outer)
+    bool bDone = false; // waits until cut is satisfied
+    while(!bDone) {
+      // we will reduce the total weight to the left of the cut over the teams
+      double storeWeightLeft = 0;
+      Kokkos::parallel_reduce(policy, KOKKOS_LAMBDA(member_type teamMember , double & weightLeft) {
+        // inner loop reduces the total weight over the threads
+        double storeTeamWeightLeft = 0;        
+        Kokkos::parallel_reduce(Kokkos::TeamThreadRange(teamMember , n_points_per_team),
+          [=] (int ii, double & teamWeightLeft) {
+          int i = ii + teamMember.league_rank() * n_points_per_team;
+          if(x(i) < cut_line) {
+            teamWeightLeft += 1.0; // for now weights are uniform: 1.0 per coordinates
+          }
+        }, storeTeamWeightLeft);
+        if(teamMember.team_rank() == 0) {
+          weightLeft += storeTeamWeightLeft;
+        }
+      }, storeWeightLeft);
+
+      // now update the cut - estimate assuming uniform weights and uniform distribution of weights
+      double storeWeightRight = static_cast<double>(n_points) - storeWeightLeft; // left + right = total
+      if(storeWeightLeft < targetWeight - epsilon) { // cut moves right
+        // the amount if weight to the right of the cut is storeWeightRight
+        // the amount of weight we want to cross shifting right is: targetWeight - storeWeightLeft
+        // the total distance on the right side is: 1.0 - cut_line    (all points are 0..1)
+        // so we assume the region from cut_line to 1.0 is uniformly weighted
+        // the fraction we want to cover is (targetWeight - storeWeightLeft) / storeWeightRight
+        // so we take that fraction of the right side distance as our estimated shift
+        cut_line += (1.0 - cut_line) * (targetWeight - storeWeightLeft) / storeWeightRight;
+      }
+      else if(storeWeightLeft > targetWeight + epsilon) { // cut moves left
+        // same as above except now we are absorbing a section on the left with total size cut_line
+        cut_line -= cut_line * (storeWeightLeft - targetWeight) / storeWeightLeft;
+      }
+      else {
+        bDone = true; // when the cut is close enough we end the main loop
+      }
+    }
+
+    // set the output values
+    Result result;
+    result.time_ms = static_cast<int>(std::chrono::duration_cast<
+      std::chrono::milliseconds>(Clock::now() - clock_start).count());
+    result.cut_line = cut_line;
+    result.n_teams = n_teams;
+    return result;
+}
+
+int main( int argc, char* argv[] )
+{
+  Kokkos::initialize(argc, argv);
```

Comment:
Prefer `Kokkos::ScopeGuard kokkosScope (argc, argv);` to initialize / finalize.

Reply:
Updated. Thanks for feedback.